packages/labs/compiler/src/lib/template-transform.ts

```diff
@@ -0,0 +1,341 @@
+      );
+    }
+
+    const f = this.context.factory;
+    const topLevelTemplates: ts.VariableStatement[] = [];
+    // For each Lit template, prepare a CompiledTemplate.
+    for (const template of this.topLevelStatementToTemplate.get(
+      node as ts.Statement
+    )!) {
+      const result = this.litHtmlPrepareRenderPhase(template.node.template);
+      if (!result.shouldCompile) {
+        throw new Error(`Unhandled TODO: Will be implemented in followup.`);
+      }
+      const {parts: partData, preparedHtml} = result;
+      const parts = createTemplateParts({f, parts: partData});
+
+      topLevelTemplates.push(
+        createCompiledTemplate({
+          f,
+          variableName: template.variableName,
+          preparedHtml,
+          parts,
+          securityBrand: this.securityBrandIdent,
+        })
+      );
+    }
+
+    const addedTopLevelTemplates: ts.Statement[] = [
+      ...(topLevelTemplates as ts.Statement[]),
+      ts.visitEachChild(
+        node,
+        (node: ts.Node) => this.rewriteTemplates(node),
+        this.context
+      ) as ts.Statement,
+    ];
+
+    // Add the security brand once.
+    if (
+      this.addedSecurityBrandVariableStatement === null &&
+      topLevelTemplates.length
+    ) {
+      this.addedSecurityBrandVariableStatement = createSecurityBrandTagFunction(
+        {
+          f,
+          securityBrandIdent: this.securityBrandIdent,
+        }
+      );
+      addedTopLevelTemplates.unshift(this.addedSecurityBrandVariableStatement);
+    }
+
+    return addedTopLevelTemplates;
+  }
+
+  /**
+   * Given the AST representation of an html tagged template expression, turn it
+   * into a CompiledTemplateResult.
+   *
+   * E.g., the following code:
+   *
+   * ```ts
+   * html`<h1>Hello ${name}${'!'}</h1>`;
+   * ```
+   *
+   * will be replaced with:
+   *
+   * ```ts
+   * { ["_$litType$"]: lit_template_1, values: [name, '!'] }
+   * ```
+   *
+   * The identifier `lit_template_1` is stored in `this.expressionToTemplate`,
+   * and references the `CompiledTemplate`.
+   *
+   * If the node passed in is not a Lit template the method returns the node
+   * unchanged.
+   */
+  private rewriteTemplateResultToCompiledTemplateResult(
+    node: ts.Node
+  ): ts.Node {
+    const f = this.context.factory;
+    if (!isLitTemplate(node)) {
+      return node;
+    }
+
+    const templateInfo = this.expressionToTemplate.get(node);
+    if (templateInfo === undefined) {
+      throw new Error(
+        `Internal Error: template info not found for ${node.getText()}`
+      );
+    }
+    return createCompiledTemplateResult({
+      f,
+      variableName: templateInfo.variableName,
+      templateExpression: node.template,
+    });
+  }
+}
+
+export const compileLitTemplates = (): ts.TransformerFactory<ts.SourceFile> =>
+  CompiledTemplatePass.getTransformer();
+
+/**
+ * E.g. html`foo` or html`foo${bar}`
+ */
+const isLitTemplate = (node: ts.Node): node is ts.TaggedTemplateExpression =>
```

Comment:
do we ever want to consider aliased imports? i.e. `import {html as h} from 'lit'`

over in the eslint plugin im considering a similar change - follow the reference rather than relying on the name alone, and check that its source `propertyName` was `html`

maybe its a far edge case though we rarely bump into and don't care?

Reply:
Great comment! As an initial goal, I've been influenced by the `minify-html-literals` rollup plugin that is being imported into the `lit` monorepo so that the two could easily be used together.

`minify-html-literals` [contains an option](https://github.com/asyncLiz/minify-html-literals#:~:text=clean%2Dcss%20defaults.-,shouldMinify,-%3F) that lets you customize which tag names are compiled. Similarly in a followup PR I'll add an option to let the user of the rollup plugin determine which templates to compile.

I think an implicit goal is also to not use any semantic type checking, and keep the initial transform purely syntactic. Thus it can work directly on JavaScript files.

The initial labs/compiler approach (once implemented) could handle your aliased import, once I've made a public API so the `h` can be detected as a template to compile.

Let me know if that should fit your use-case and thank you for the great comment!

Reply:
very good point, i may actually do the same in the lint plugin in that case (configurable identifier)

thanks for the explanation

Reply:
I think this compiler has a higher correctness bar than a minifier or linter.

This transform will definitely break any tagged template literal that isn't specifically the Lit html TTL, so I think we need to verify that it's imported from either 'lit' or 'lit-html' or 'lit-element' (and not, I think, 'lit/static.js' and friends, let alone other HTML TTL vendors, like preact)

Reply:
If we're going to handle this in a future change, let's leave a comment about it

Reply:
Comment added. Good call.

Reply:
I'm exploring a technique that's not too complicated, but still allows some flexibility in the `html` import in https://github.com/lit/lit/pull/4055. It did end up requiring use of the TypeChecker.

We will still probably need to vend a way for a consumer to customize what module specifiers are valid exports of `html`.